### DISCUSSION TITLE
Confused with returning Ref<T> as a type to a Pinia state

### DISCUSSION THREAD

**Discussion by philippedasilva-orizone** [vuejs/pinia#485] [Help and Questions] [answered]
Hi,

I'm slightly confused with Pinia and Typescript...

I am creating a simple Authentication store where I'm initializing my Pinia store state using `useStorage` from VueUse so that I don't lose my user session when reloading the page.
Note that the code below actually executes perfectly while I believe it shouldn't...

```typescript
import firebase from 'firebase/app';
import 'firebase/auth';
import { defineStore } from 'pinia';
import { useStorage } from '@vueuse/core';
import { router } from '../routes';
import { Authentication } from 'origami-common';

export const useAuthenticationStore = defineStore({
    id: "authentication",
    state() {
        return useStorage<Authentication.UserProfile>(
            "user",
            { accessToken: "", displayName: "", photoUrl: "", emailAddress: "" },
            localStorage);
    },
    getters: {
        isAuthenticated: (state) => state.accessToken !== "",
        getToken: (state) => state.accessToken
    },
    actions: {
        async signInWithGoogle() {
            const provider = new firebase.auth.GoogleAuthProvider();
            provider.addScope("https://mail.google.com");
            try {
                const signInResponse = await firebase.auth().signInWithPopup(provider);
                const credential = signInResponse.credential as firebase.auth.OAuthCredential;
                this.accessToken = credential.accessToken as string;
                this.displayName = signInResponse.user?.displayName as string;
                this.photoUrl = signInResponse.user?.photoURL as string;
                this.emailAddress = signInResponse.user?.email as string;
                router.push("/");
            } catch (error) {
                console.error(error);
            }
        },
        async signOut() {
            try {
                await firebase.auth().signOut();
                this.accessToken = "";
                this.displayName = "";
                this.photoUrl = "";
                this.emailAddress = "";
                router.push("/signin")
            } catch (error) {
                console.error(error);
            }
        }
    }
});
```

Really nothing rocket science there.
Still, the type returned by `useStorage` is of type Ref<Authentication.UserProfile> but on vscode, intellisense tells me that the `state` parameter passed to getters is of type `any` and in actions, `this` is of type Ref<Authentication.UserProfile>.

![image](https://user-images.githubusercontent.com/77384971/117580671-5ab3a500-b0f9-11eb-9498-cb9aaf5b05e6.png)
![image](https://user-images.githubusercontent.com/77384971/117580692-6dc67500-b0f9-11eb-8a04-77fe523038d0.png)

I would actually expect to make calls in actions such as `this.value.accessToken = ...` but if I do so, it just doesn't work... I get an error saying `value` is undefined.

The only way for me to actually get no errors is to set the return type of the `useStorage` function `as unknown` and then casting it to `as Authentication.UserProfile` this way:
```
export const useAuthenticationStore = defineStore({
    id: "authentication",
    state() {
        return (useStorage<Authentication.UserProfile>(
            "user",
            { accessToken: "", displayName: "", photoUrl: "", emailAddress: "" },
            localStorage) as unknown) as Authentication.UserProfile;
    },
    getters: {
        isAuthenticated: (state) => state.accessToken !== "",
        getToken: (state) => state.accessToken
    },
```

Then, I see no more intellisense typescript errors.
However, the type of the `state.accessToken` argument passed to getters is still set to `any`
![image](https://user-images.githubusercontent.com/77384971/117580896-6b184f80-b0fa-11eb-8c5c-cbf1782736c7.png)

I'm using Vue 3.0.5, Pinia 2.0.0.alpha16 , typescript 4.2.4 and vitejs 2.2.4

**Comment by posva** [comment, marked as answer]
The state should return a plain object (it can be a reactive object as well). It cannot be a Ref. So maybe do:

```ts
state() {
        return {
           user: useStorage<Authentication.UserProfile>(
            "user",
            { accessToken: "", displayName: "", photoUrl: "", emailAddress: "" },
            localStorage);
  }
    },
```

The problem is that from a typing perspective, it will think the type of `this.user` is a ref while in practice it isn't. I think we could unwrap refs automatically in pinia types

**Reply by philippedasilva-orizone** [reply]
As you stated, if I do what you suggest, it thinks the type of `this.user` is `Ref<Authentication.UserProfile>`and I still get no typing on the state in the getters.

What do you mean by "unwrapping refs automatically"? Using `unref()` in pinia types?

**Reply by philippedasilva-orizone** [reply]
I did experiment with `unref()` and it seems to work just fine. 
```typescript
export const useAuthenticationStore = defineStore({
  id: 'authentication',
  state: () =>
    unref(
      useStorage<Authentication.UserProfile>(
        'user',
        {
          accessToken: '',
          displayName: '',
          photoUrl: '',
          emailAddress: '',
        },
        localStorage
      )
    ),
// rest of the code remains unchanged...
```

![image](https://user-images.githubusercontent.com/77384971/117657001-563ec900-b199-11eb-8aa7-8a1df692e9b4.png)

If I don't see other subsequent issues, I'll dig into pinia state registration code and see if I can do something along these lines.

**Reply by posva** [reply]
Ideally this should be supported natively: https://github.com/posva/pinia/issues/491

The workaround currently is doing an `as MyType` at the end of `useStorage()` to force the type inside state

**Reply by philippedasilva-orizone** [reply]
@posva I'm using `unref()` as seen in my previous post and it's working fine for me with no type errors. Should I still go for `as MyType` anyway?

**Reply by posva** [reply]
Yes, because unref will break the reactivity

**Reply by philippedasilva-orizone** [reply]
I will then until I (or we) get this implemented automatically in pinia ;)
I guess, I'll have to live with this annoying type error in vs code for the time being...

![image](https://user-images.githubusercontent.com/77384971/117737925-27593f00-b1fb-11eb-84db-9ed84bad1357.png)
![image](https://user-images.githubusercontent.com/77384971/117737981-4788fe00-b1fb-11eb-8ff8-3e6d5488fd62.png)

It's not like if the code wasn't running after all :p

**Reply by posva** [reply]
the error tells you what to do to remove it: `as unknown as MyType`

**Reply by philippedasilva-orizone** [reply]
Yes and forgive being new to typescript (while pretty familiar with strongly typed languages such as C++ or C#) but it feels weird having to cast some variable to `unknown` and then to the actual type it is supposed to have... Isn't it? or is it just me? :)
